src/mscorlib/src/System/Span.cs

```diff
@@ -164,5 +167,145 @@ public bool Equals(Span<T> other)
+            SpanContracts.RequiresInInclusiveRange(values._length, _length);
+
+            SpanHelper.CopyTo<T>(ref _rawPointer, ref values._rawPointer, values.Length);
+        }
+
+        public void Set(T[] values)
+        {
+            Set(new ReadOnlySpan<T>(values));
+        }
+
+        [CLSCompliant(false)]
+        public unsafe void Set(void* source, int elementsCount)
+        {
+            SpanContracts.RequresNoReferences<T>();
+            SpanContracts.RequiresNonNegative(elementsCount);
+            SpanContracts.RequiresInInclusiveRange(_length, elementsCount);
+
+            SpanHelper.Memmove<T>((byte*)_rawPointer, (byte*)source, elementsCount);
+        }
+    }
+
+    internal static class SpanContracts
+    {
+        internal static void RequiresIndexInRange(int index, int length) {
+            if ((uint)index >= (uint)length)
+                ThrowHelper.ThrowIndexOutOfRangeException();
+        }
+
+        internal static void RequiresInRange(int start, int length) {
+            if ((uint)start >= (uint)length)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+        }
+
+        public static void RequiresInInclusiveRange(int start, int length) {
+            if ((uint)start > (uint)length)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+        }
+
+        public static void RequiresNonNegative(int number) {
+            if (number < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+        }
+
+        internal static void RequiresNonNullArray<T>(T[] array) {
+            if (array == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+        }
+
+        internal static void RequiresNoCovariance<T>(T[] array) {
+            if (default(T) == null) { // Arrays of valuetypes are never covariant
+                if (array.GetType() != typeof(T[]))
+                    ThrowHelper.ThrowArrayTypeMismatchException();
+            }
+        }
+
+        /// <summary>
+        /// throws for reference types and value types with references, they must not be put to unmanaged memory
+        /// </summary>
+        internal static void RequresNoReferences<T>() {
+            if (JitHelpers.ContainsReferences<T>())
+                ThrowHelper.ThrowInvalidTypeForUnmanagedMemory(typeof(T));
+        }
+    }
+
+    internal static class SpanHelper
+    {
+        internal static void CopyTo<T>(ref IntPtr destination, ref IntPtr source, int elementsCount)
+        {
+            if (!JitHelpers.ContainsReferences<T>())
+            {
+                unsafe
+                {
+                    Memmove<T>((byte*)destination, (byte*)source, elementsCount);
+                }
+            }
+            else
+            {
+                ref T dest = ref JitHelpers.GetByRef<T>(ref destination);
+                ref T src = ref JitHelpers.GetByRef<T>(ref source);
+
+                for (int i = 0; i < elementsCount; i++)
```

Comment:
It seems like we should at least fail if the destination is within the source buffer.